enhancements/baremetal/baremetal-ironic-config-passthrough.md

```diff
@@ -0,0 +1,140 @@
+---
+title: baremetal-ironic-config-passthrough
+authors:
+  - "@derekh"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-07-17
+last-updated: 2020-07-17
+status: provisional
+see-also:
+  - "https://github.com/openshift/installer/pull/3887"
+  - "https://github.com/openshift/machine-api-operator/pull/645"
+---
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Baremetal IPI makes use of OpenStack Ironic to provision RHCOS to each baremetal
+host. Ironic is deployed in the metal3 pod and its configuration is highly opinionated.
+This enhancement would provide a mechanism to pass arbitary ironic config options
+into the metal3 pod in order to change how ironic and ironic-inspector are configured.
+
+## Motivation
+
+The current opinionated ironic works well when we have a single set of target hardware
+to support, this is the ideal but hasn't been our experience. As users deploy IPI clusters
+in labs, production systems, and for CI, we occasionally needs to deal with corner cases arise.
+
+The motivation for this enhancement is to allow people to deal with their individual corner
+cases without the need to insert special cases into the ironic containers. The ironic
+container can stay opinionated and targeted for a specific set of environments while also
+being usable where uniq envirnments require adjustments.
+
+In addition, when a change is being considered to ironic's opinionated configurations,
+deployers will be able to test this change in their deployment and make recommendations on
+an the suitability of making a config update permanent in the ironic image.
+
+### Goals
+
+Allow users of to tweak ironic config options on the fly without the need for a new
+ironic image.
+
+### Non-Goals
+
+Our goal is not to provide an API the ends up being commonly used with a small
+subset of options. i.e. if any specific config option becomes a frequently
+tweaked option for multiple environments then it should be considered a condidate
+as a top-level platform configuration.
+
+## Proposal
+
+Add a new platform configuration "ironicExtraConf" of type map, this would consist
+of keys and values to be used as config options for configuring ironic.
+
+Each config option should be expressed as a key/value pair with the format
+
+OS_<section>_\_<name>=<value> - where `section` and `name` are the
+reprepresent the config option in ironic.conf e.g. to set a IPA ssh key and
+set the number of ironic API workers
+
+```yaml
+platform:
+  baremetal:
+    ironicExtraConf: {"OS_PXE__PXE_APPEND_PARAMS":'nofb nomodeset vga=normal sshkey="ssh-rsa AAAA..."', "OS_API__API_WORKERS":"8"}
```

Comment:
Ironic (and IPA) was always intended to be an implementation detail for IPI, so I'm pretty uncomfortable with adding a top-level ironicFoo option tbh.

In particular if you check the discussion on https://github.com/coreos/ignition/issues/979 you'll see there are a bunch of unsolved issues wrt the OpenShift coreos-install based ramdisk vs IPA, and the options in either case are completely different.

I wonder if we could perhaps make this more generic and have some pxeExtra interface or something, or even better some interface which describes the required behavior e.g which console to use?

Reply:
Yes, upstream we want to use the oslo.config environment variable interface instead of customizing individual settings in the ironic config to cut down on the number of changes to the image. The whole point of adding environment variable handling to oslo.config was to make it easier to pass settings to OpenStack services running in containers.

Downstream in OpenShift, however, I would rather that we add explicit APIs for the things we need to allow users to change. That allows us to take advantage of the API server for some basic validation, and makes it harder for users to put the system into an unsupportable state or one where metal3 is crash-looping due to invalid settings.